lumen/ai/agents.py

```diff
@@ -723,6 +725,183 @@ async def respond(
+
+        system_prompt = await self._render_prompt(
+            "main",
+            messages,
+            has_errors=bool(errors),
+        )
+
+        with self._add_step(title=title or "dbt Semantic Layer query", steps_layout=self._steps_layout) as step:
+            model_spec = self.prompts["main"].get("llm_spec", self.llm_spec_key)
+            response = self.llm.stream(
+                messages,
+                system=system_prompt,
+                model_spec=model_spec,
+                response_model=DbtslQueryParams,
+            )
+
+            query_params = None
+            try:
+                async for output in response:
+                    step_message = output.chain_of_thought or ""
+                    step.stream(step_message, replace=True)
+
+                query_params = {
+                    "metrics": output.metrics,
+                    "group_by": output.group_by,
+                    "limit": output.limit,
+                    "order_by": output.order_by,
+                    "where": output.where,
+                }
+                expr_slug = output.expr_slug
+
+                # Ensure required fields are present
+                if "metrics" not in query_params or not query_params["metrics"]:
+                    raise ValueError("Query must include at least one metric")
+
+                # Ensure limit exists with a default value if not provided
+                if "limit" not in query_params or not query_params["limit"]:
+                    query_params["limit"] = 100
+
+                formatted_params = json.dumps(query_params, indent=2)
+                step.stream(f"\n\n`{expr_slug}`\n```json\n{formatted_params}\n```")
+
+                self._memory["dbtsl_query_params"] = query_params
+            except asyncio.CancelledError as e:
+                step.failed_title = "Cancelled dbt Semantic Layer query generation"
+                raise e
+
+        if step.failed_title and step.failed_title.startswith("Cancelled"):
+            raise asyncio.CancelledError()
+        elif not query_params:
+            raise ValueError("No dbt Semantic Layer query was generated.")
+
+        try:
+            # Execute the query against the dbt Semantic Layer
+            client = self._get_dbtsl_client()
+            async with client.session():
+                sql_query = await client.compile_sql(
+                    metrics=query_params.get('metrics', []),
+                    group_by=query_params.get('group_by'),
+                    limit=query_params.get('limit'),
+                    order_by=query_params.get('order_by'),
+                    where=query_params.get('where')
+                )
+
+                # Log the compiled SQL for debugging
+                step.stream(f"\nCompiled SQL:\n```sql\n{sql_query}\n```", replace=False)
+
+            # Create a SQL source from the compiled query
+            source = self._memory["source"]
```

Comment:
Not a totally safe assumption that the `source` is the same source that you can execute the Dbt generated queries against. Seems like we might have to make source a parameter on the agent in this case.